server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java

```diff
@@ -0,0 +1,488 @@
+            if (nodeEntry.hasData() == false && nodeEntry.isFetching() == false) {
+                nodesToFetch.add(nodeEntry);
+            }
+        }
+        return nodesToFetch;
+    }
+
+    /**
+     * Are there any nodes that are fetching data?
+     */
+    private boolean hasAnyNodeFetching(Map<String, AsyncBatchShardFetch.NodeEntry<T>> shardCache) {
+        for (AsyncBatchShardFetch.NodeEntry<T> nodeEntry : shardCache.values()) {
+            if (nodeEntry.isFetching()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Async fetches data for the provided shard with the set of nodes that need to be fetched from.
+     */
+    // visible for testing
+    void asyncFetch(final DiscoveryNode[] nodes, long fetchingRound) {
+        logger.trace("{} fetching [{}] from {}", batchUUID, type, nodes);
+        action.list(nodes, shardsToCustomDataPathMap, new ActionListener<BaseNodesResponse<T>>() {
+            @Override
+            public void onResponse(BaseNodesResponse<T> response) {
+                processAsyncFetch(response.getNodes(), response.failures(), fetchingRound);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                List<FailedNodeException> failures = new ArrayList<>(nodes.length);
+                for (final DiscoveryNode node : nodes) {
+                    failures.add(new FailedNodeException(node.getId(), "total failure in fetching", e));
+                }
+                processAsyncFetch(null, failures, fetchingRound);
+            }
+        });
+    }
+
+    /**
+     * The result of a fetch operation. Make sure to first check {@link #hasData()} before
+     * fetching the actual data.
+     */
+    public static class FetchResult<T extends BaseNodeResponse> {
+
+        private final Map<DiscoveryNode, T> data;
+        private Map<ShardId, Set<String>> ignoredShardToNodes = new HashMap<>();
+
+        public FetchResult(Map<DiscoveryNode, T> data, Map<ShardId, Set<String>> ignoreNodes) {
+            this.data = data;
+            this.ignoredShardToNodes = ignoreNodes;
+        }
+
+        /**
+         * Does the result actually contain data? If not, then there are on going fetch
+         * operations happening, and it should wait for it.
+         */
+        public boolean hasData() {
+            return data != null;
+        }
+
+        /**
+         * Returns the actual data, note, make sure to check {@link #hasData()} first and
+         * only use this when there is an actual data.
+         */
+        public Map<DiscoveryNode, T> getData() {
+            assert data != null : "getData should only be called if there is data to be fetched, please check hasData first";
+            return this.data;
+        }
+
+        /**
+         * Process any changes needed to the allocation based on this fetch result.
+         */
+        public void processAllocation(RoutingAllocation allocation) {
```

Comment:
Same as `AsyncShardFetch` with difference now we are adding ignorenodes for a set of shards